### ISSUE TITLE
FR: Export highlights - adjustable file name format

### ISSUE THREAD

**Issue by mergen3107** [koreader/koreader#12587]
**Does your feature request involve difficulty completing a task? Please describe.**
I would love to be able to change the filename format for exported highlights.

**Describe the solution you'd like**
Something like in Screensaver message plugin, which suppost variables with `%`.
Say, I would love to have somthing like this:
`yyyy-mm-dd Author Name - Book Title.txt`

Currently, it does this:
`yyyy-mm-dd-hh-mm-ss-Book Title.txt`

**Describe alternatives you've considered**
I can of course edit the format in the plugin code itself, but would love to have the option to modify it from plugin settings for these reasons:
- I'd need to keep updating my modified plugin folder after plugin updates
- Not user-firendly, whereas I would like other people to enjoy this FR too
- Not a good habit of modifying plugins :D 

**Additional context**
Otherwise exporter plugin is great!

**Comment by mergen3107** [comment]
Yeah, my suggested format above also implies that we should be able to overwrite the existing file. A new dialog is required?

**Comment by hius07** [comment]
Userpatch, overwrites existing file silently.
```Lua
local userpatch = require("userpatch")
local util = require("util")
userpatch.registerPatchPluginFunc("exporter", function(plugin)
    local init_orig = plugin.init
    plugin.init = function(self)
        init_orig(self)
        for _, v in pairs(self.targets) do
            v.getFilePath = function(v_self, t)
                if self.is_remote then return end
                local plugin_settings = G_reader_settings:readSetting("exporter") or {}
                local clipping_dir = plugin_settings.clipping_dir or v_self.clipping_dir
                local filename
                if #t == 1 then
                    if plugin_settings.clipping_dir_book then
                        clipping_dir = util.splitFilePathName(t[1].file):sub(1, -2)
                    end
                    -- single book export filename
                    filename = string.format("%s %s - %s.%s", os.date("%Y-%m-%d"), t[1].author, t[1].title, v_self.extension)
                else
                    local title = v_self.all_books_title or "all-books"
                    filename = string.format("%s-%s.%s", v_self:getTimeStamp(), title, v_self.extension)
                end
                return clipping_dir .. "/" .. util.getSafeFilename(filename)
            end
        end
    end
end)
```

**Comment by mergen3107** [comment]
Thanks!

I did:
- make a new file `2-exporter-file-format.lua` and put it in `patches` folder
- restart KOReader - no errors about the patch, it shows up in the patch list OK, as enabled.
- Exported notes again - but it still uses the same format.

It seems it didn't work.

How should I run this?

**Comment by mergen3107** [comment]
I see these lines in `screensaver.lua`:
https://github.com/koreader/koreader/blob/f0e5acfd263dc9245483d7f7627676dbd06fadbb/frontend/ui/screensaver.lua#L108-L118

Can at least `author` be visible to the exporter plugin as well?
So we can re-write this part of the `filename` function to be selectable from the menu?
https://github.com/koreader/koreader/blob/f0e5acfd263dc9245483d7f7627676dbd06fadbb/plugins/exporter.koplugin/base.lua#L106

**Comment by hius07** [comment]
> * make a new file `2-exporter-file-format.lua` and put it in `patches` folder

Yes, correct.
Tested, it should work for current book exporting. What export format do you use? Anything in the crash.log near the line "applying patch"?

**Comment by mergen3107** [comment]
> What export format do you use?
.txt

I will check the crash.log a little later

**Comment by mergen3107** [comment]
Interesting, this time it worked!
Thank you again!

I wonder if it has something to do with this other patch: https://github.com/koreader/koreader/issues/12589#issuecomment-2394877242

Let me try again with enabling it

**Comment by mergen3107** [comment]
When it worked, the lines around the applying patch were just:
```
10/04/24-23:22:43 INFO  Applying patch: ./patches/2-exporter-file-format.lua 
10/04/24-23:22:43 DEBUG userpatch registered for plugin exporter 
```

Full log: 
[exporter_crashlog.txt](https://github.com/user-attachments/files/17265029/exporter_crashlog.txt)
- Restart KOReader
- Find that book
- Export notes - popup was correct as well
- Exit

**Comment by mergen3107** [comment]
Nope, it works either way.

I honestly don't know why it didn't work the first time O_O

**Comment by mergen3107** [comment]
Works great! (Will come back if Exporter gets updated).

**Comment by cameronj86** [comment]
For the life of me, I can't get the tweak to work. 

Everything is the same as the the earlier solution, except I've changed `filename` to be `Author - Date.extension`:
```lua
    -- single book export filename
                        filename = string.format("%s - %s.%s", t[1].title, os.date("%Y-%m-%d"), v_self.extension)
else
    local title = v_self.all_books_title or "all-books"
    filename = string.format("%s-%s.%s", title, v_self:getTimeStamp(), v_self.extension)
```

My patch is listed as enabled under `Patch management` but exported files never translates the file. `Patches executed = 0`if that matters. 

Curious if anyone has any ideas? No other patches are used by me

**Comment by hius07** [comment]
After https://github.com/koreader/koreader/pull/12641 the patch doesn't work anymore.

**Comment by hius07** [comment]
Reopening until a fix or a userpatch.

**Comment by hius07** [comment]
Ah, sorry, false alert, all works well, and your changed line works too.
I'm on the latest nightly build.
